src/node_http2.cc

```diff
@@ -912,38 +769,33 @@ void Http2Session::OnHeaders(Nghttp2Stream* stream,
 
   Isolate* isolate = env()->isolate();
   HandleScope scope(isolate);
-  Local<Object> holder = Object::New(isolate);
-  holder->SetPrototype(context, v8::Null(isolate)).ToChecked();
   Local<String> name_str;
   Local<String> value_str;
-  Local<Array> array;
-  while (headers != nullptr) {
-    nghttp2_header_list* item = headers;
-    name_str = ExternalHeaderNameResource::New(isolate, item->name);
-    nghttp2_vec name = nghttp2_rcbuf_get_buf(item->name);
-    nghttp2_vec value = nghttp2_rcbuf_get_buf(item->value);
-    value_str = String::NewFromUtf8(isolate,
-                                    reinterpret_cast<char*>(value.base),
-                                    v8::NewStringType::kNormal,
-                                    value.len).ToLocalChecked();
-    if (holder->Has(context, name_str).FromJust()) {
-      if (CheckHeaderAllowsMultiple(&name)) {
-        Local<Value> existing = holder->Get(context, name_str).ToLocalChecked();
-        if (existing->IsArray()) {
-          array = existing.As<Array>();
-          array->Set(context, array->Length(), value_str).FromJust();
-        } else {
-          array = Array::New(isolate, 2);
-          array->Set(context, 0, existing).FromJust();
-          array->Set(context, 1, value_str).FromJust();
-          holder->Set(context, name_str, array).FromJust();
-        }
-      }  // Ignore singleton headers that appear more than once
-    } else {
-      holder->Set(context, name_str, value_str).FromJust();
+
+  // TODO(jasnell): get the size of the array
+  Local<Array> holder = Array::New(isolate);
+  Local<Function> fn = env()->push_values_to_array_function();
+  Local<Value> argv[NODE_PUSH_VAL_TO_ARRAY_MAX * 2];
+
+  do {
+    size_t j = 0;
+    while (headers != nullptr && j < arraysize(argv) / 2) {
+      nghttp2_header_list* item = headers;
+      name_str = ExternalHeaderNameResource::New(isolate, item->name);
+      nghttp2_vec value = nghttp2_rcbuf_get_buf(item->value);
+      value_str = String::NewFromUtf8(isolate,
+                                      reinterpret_cast<char*>(value.base),
+                                      v8::NewStringType::kNormal,
+                                      value.len).ToLocalChecked();
+      argv[j * 2] = name_str;
+      argv[j * 2 + 1] = value_str;
+      headers = item->next;
+      j++;
     }
-    headers = item->next;
-  }
+    if (j > 0) {
+      fn->Call(env()->context(), holder, j * 2, argv).ToLocalChecked();
+    }
+  } while (headers != nullptr);
```

Comment:
can you add some comments explaining how all of this works? It's a bit obscure and it might help in the future.